diff -uprN a/eeprom.c b/eeprom.c
--- a/eeprom.c	2020-04-04 17:56:49.177167014 +0800
+++ b/eeprom.c	2020-04-04 17:58:18.189565815 +0800
@@ -80,13 +80,14 @@ mt76_eeprom_override(struct mt76_dev *de
 	const u8 *mac;
 
 	if (!np)
-		return;
+		goto out;
 
 	mac = of_get_mac_address(np);
 	if (!IS_ERR_OR_NULL(mac))
 		memcpy(dev->macaddr, mac, ETH_ALEN);
 #endif
 
+out:
 	if (!is_valid_ether_addr(dev->macaddr)) {
 		eth_random_addr(dev->macaddr);
 		dev_info(dev->dev,
